File: code_snippets/python-sdk/Search/as/pagination.py
from clarifai.client.user import User
#replace your "user_id"
client = User(user_id="user_id")

#(Example Workflows: 'Universal', 'Empty', 'General')
app = client.create_app(app_id="app_id", base_workflow="Universal")

# set the input url's
urls = [
    "https://images.pexels.com/photos/139257/pexels-photo-139257.jpeg",
    "https://images.pexels.com/photos/1879386/pexels-photo-1879386.jpeg",
    "https://images.pexels.com/photos/1071882/pexels-photo-1071882.jpeg"
]

input_obj = app.inputs()
# add inputs to the app
for i, url in enumerate(urls):
    input_obj.upload_from_url(input_id=f"input{i}", image_url=url)


# perform search with pagination
search = app.search(pagination=True)
response = search.query(ranks=[{"text_raw": "Red pineapples on the beach."}], per_page=2,page_no=1)
resp = list(response)
print(len(resp[0].hits))

File: code_snippets/python-sdk/Search/as/pagination.ts
import { Input, Search, User } from "clarifai-nodejs";


const appId = "test_app"; // Placeholder for application ID

// Initialize a User object with your credentials
const client = new User({
  userId: process.env.CLARIFAI_USER_ID,
  pat: process.env.CLARIFAI_PAT,
  appId,
});

// Create an application with the specified app ID and base workflow
await client.createApp({ appId, baseWorkflow: "Universal" });

// URLs of images to be uploaded
const urls = [
  "https://images.pexels.com/photos/139257/pexels-photo-139257.jpeg",
  "https://images.pexels.com/photos/1879386/pexels-photo-1879386.jpeg",
  "https://images.pexels.com/photos/1071882/pexels-photo-1071882.jpeg",
];

// Initialize an Input object to manage input data
const input = new Input({
  authConfig: {
    userId: process.env.CLARIFAI_USER_ID,
    pat: process.env.CLARIFAI_PAT,
    appId,
  },
});

// Upload each image from the provided URLs
for (let i = 0; i < urls.length; i++) {
  await input.uploadFromUrl({ inputId: `input${i}`, imageUrl: urls[i] });
}

// Initialize a search object for the created application, top-K not set for pagination
const search = new Search({
  authConfig: {
    userId: process.env.CLARIFAI_USER_ID,
    pat: process.env.CLARIFAI_PAT,
    appId,
  },
  metric: "euclidean",
});

const response = search.query({
  ranks: [{ textRaw: "Red pineapples on the beach." }],
  perPage: 2,
  page: 1,
});

const hits = (await response.next()).value?.hitsList ?? [];
console.log(hits);
